Mathlib/Algebra/Lie/Weights/Chain.lean

```diff
@@ -217,4 +217,124 @@ end IsCartanSubalgebra
+  by_cases hα : α = 0
+  · rw [hα, smul_zero, zero_add]; exact β.weightSpace_ne_bot
+  classical
+  rw [← Nat.lt_succ, Nat.succ_eq_add_one, chainTopCoeff_add_one _ _ hα] at hn
+  exact Nat.find_min (eventually_weightSpace_smul_add_eq_bot M α β hα).exists hn
+
+lemma weightSpace_chainBotCoeff_sub_one_zsmul_sub :
+    weightSpace M ((-chainBotCoeff α β - 1 : ℤ) • α + β : L → R) = ⊥ := by
+  rw [sub_eq_add_neg, ← neg_add, neg_smul, ← smul_neg, chainBotCoeff,
+    weightSpace_chainTopCoeff_add_one_zsmul_add _ _ (by simpa using hα)]
+
+lemma weightSpace_zsmul_add_ne_bot {n : ℤ}
+    (hn : -chainBotCoeff α β ≤ n) (hn' : n ≤ chainTopCoeff α β) :
+      weightSpace M (n • α + β : L → R) ≠ ⊥ := by
+  rcases n with (n | n)
+  · simp only [Int.ofNat_eq_coe, Nat.cast_le, ← nsmul_eq_smul_cast] at hn' ⊢
+    exact weightSpace_nsmul_add_ne_bot_of_le α β hn'
+  · simp only [Int.negSucc_eq, ← Nat.cast_succ, neg_le_neg_iff, Nat.cast_le] at hn ⊢
+    rw [neg_smul, ← smul_neg, ← nsmul_eq_smul_cast]
+    exact weightSpace_nsmul_add_ne_bot_of_le (-α) β hn
+
+lemma weightSpace_neg_zsmul_add_ne_bot {n : ℕ} (hn : n ≤ chainBotCoeff α β) :
+    weightSpace M ((-n : ℤ) • α + β : L → R) ≠ ⊥ := by
+  apply weightSpace_zsmul_add_ne_bot α β <;> omega
+
+/-- This is the first weight in an `α`-chain through `β`. -/
+noncomputable
+def chainTop (α : L → R) (β : Weight R L M) : Weight R L M :=
+  ⟨chainTopCoeff α β • α + β, weightSpace_nsmul_add_ne_bot_of_le α β le_rfl⟩
+
+/-- This is the first weight in an `α`-chain through `β`. -/
+noncomputable
+def chainBot (α : L → R) (β : Weight R L M) : Weight R L M :=
+  ⟨(- chainBotCoeff α β : ℤ) • α + β, weightSpace_neg_zsmul_add_ne_bot α β le_rfl⟩
+
+lemma coe_chainTop' : (chainTop α β : L → R) = chainTopCoeff α β • α + β := rfl
+
+@[simp] lemma coe_chainTop : (chainTop α β : L → R) = (chainTopCoeff α β : ℤ) • α + β := by
+  rw [← nsmul_eq_smul_cast ℤ]; rfl
+@[simp] lemma coe_chainBot : (chainBot α β : L → R) = (-chainBotCoeff α β : ℤ) • α + β := rfl
+
+@[simp] lemma chainTop_neg : chainTop (-α) β = chainBot α β := by ext; simp
+@[simp] lemma chainBot_neg : chainBot (-α) β = chainTop α β := by ext; simp
+
+@[simp] lemma chainTop_zero : chainTop 0 β = β := by ext; simp
+@[simp] lemma chainBot_zero : chainBot 0 β = β := by ext; simp
+
+lemma weightSpace_add_chainTop :
+    weightSpace M (α + chainTop α β : L → R) = ⊥ := by
+  rw [coe_chainTop', ← add_assoc, ← succ_nsmul', weightSpace_chainTopCoeff_add_one_nsmul_add _ _ hα]
+
+lemma weightSpace_neg_add_chainTop :
+    weightSpace M (-α + chainBot α β : L → R) = ⊥ := by
+  rw [← chainTop_neg, weightSpace_add_chainTop _ _ (by simpa using hα)]
+
+lemma chainTop_isNonZero' (hα' : weightSpace M α ≠ ⊥) :
+    (chainTop α β).IsNonZero := by
+  by_contra e
+  apply hα'
+  rw [← add_zero (α : L → R), ← e, weightSpace_add_chainTop _ _ hα]
+
+lemma chainTop_isNonZero (α β : Weight R L M) (hα : α.IsNonZero) :
+    (chainTop α β).IsNonZero :=
+  chainTop_isNonZero' α β hα α.2
```

Comment:
Should we have analogous lemmas for `chainBot`?

Reply:
I don't think this is possible in the general case. For example `chainBot α α` will be zero if `-α` is not a weight (or will this never happen)?

Reply:
Fair enough

bors d+